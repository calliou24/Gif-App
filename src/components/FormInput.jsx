import React, {useState} from 'react';
import PropTypes from 'prop-types'



export const FormInput = ( {add, reset} ) => {

    const [targetValue , setTargetValue] = useState('')
    //important that this use state have the argument of an Empty String of something 
    //we dont want to have a undefined props inside our component
    function sendValue(e) {
        
        e.preventDefault()
        add( e.target.firstChild.value)
        e.target.firstChild.value = ""
        setTargetValue("")
    }
    function setTargetValueReset(e) {
        
        e.preventDefault()
        reset()

    }
    return (
        <div className='form-cont'> 
        <p>Search</p>
        <form  onSubmit={ e => sendValue( e )}>
            <input onChange={ e => setTargetValue( e.target.value)}placeholder="..."/>
            <p className="text-reflect">{targetValue}</p>
            <button type='submit' className='btn-add-cat'>Add</button>
            <button onClick={e => setTargetValueReset(e)} type='submit' className='btn-add-cat'>Reset</button>
        </form>
        </div>
    )
} 

FormInput.propTypes = { 
    add : PropTypes.func.isRequired,
    reset : PropTypes.func.isRequired
}