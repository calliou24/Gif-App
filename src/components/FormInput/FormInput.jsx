import React, { useState } from "react";
import PropTypes from "prop-types";

import styles from "./form.module.css";

export const FormInput = ({ add, reset }) => {
  const [targetValue, setTargetValue] = useState("");
  //important that this use state have the argument of an Empty String of something
  //we dont want to have a undefined props inside our component

  function setValue(e) {
    setTargetValue(e.target.value);
  }

  function sendValue(e) {
    e.preventDefault();
    add(targetValue);
    setTargetValue("");
  }

  function setTargetValueReset(e) {
    e.preventDefault();
    reset();
  }
  return (
    <form className={styles.form} onSubmit={(e) => sendValue(e)}>
      <div className={styles.inputCont}>
        <input
          className={styles.searchBar}
          onChange={(e) => setValue(e)}
          placeholder="Search"
          value={targetValue}
        />
        <p id="p" className={styles.textReflect}>
          {targetValue.length < 20 ? (
            targetValue
          ) : (
            <p style={{ color: "red" }}>Too many characters!</p>
          )}
        </p>
      </div>
      <div className={styles.btnCont}>
        <button type="submit" className={styles.btn}>
          ADD
        </button>
        <button
          onClick={(e) => setTargetValueReset(e)}
          type="submit"
          className={styles.btn}
        >
          DEFAULT
        </button>
      </div>
    </form>
  );
};

FormInput.propTypes = {
  add: PropTypes.func.isRequired,
  reset: PropTypes.func.isRequired,
};
