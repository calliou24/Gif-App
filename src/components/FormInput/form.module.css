.form {
  display: flex;
  align-items: center;
}

.inputCont {
  position: relative;
}

.searchBar {
  border: none;
  outline: none;
  font-size: 1.2rem;
  background-color: var(--secondary-color);
  padding: 15px;
}

.textReflect {
  color: var(--text-reflect);
  font-size: 1.3rem;
  bottom: -30px;
  left: 0;
  position: absolute;
}

.btnCont {
  display: flex;
  align-items: center;
  column-gap: 15px;
  margin-left: 15px;
}
.btn {
  border: none;
  outline: none;
  font-size: 1.2rem;
  background-color: var(--secondary-color);
  padding: 15px;
  cursor: pointer;
}

@media (max-width: 50em) {
  .form {
    flex-direction: column;
    align-items: flex-start;
    row-gap: 50px;
  }
  .btnCont {
    margin-left: 0;
  }

  .btn {
    font-size: 1rem;
    padding: 10px;
  }
  .searchBar {
    font-size: 1rem;
  }
}
