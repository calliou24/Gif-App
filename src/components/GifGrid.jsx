import React,{ useState, useEffect } from 'react'
import { GifComponent } from './GifComponent/GifComponent'
import { useFetchGifs } from '../hooks/useFetchGifs'
export const GifGrid = ({category}) => {

  const {data, loading} = useFetchGifs( category ) 
    //use effect nos permite ejecutar algo una sola vez al renderizar el componente
    //de manera que este no esta cambiando y haciendo la funcion cada que se actualizar 
    //por ejemplo en un useState, entrando en un bucle
    //la funcion que se necestia aqui con el nombre getData esta almacenada en una carpeta
    //para poder modularizar nuestro archivo
  return (
    <div className='gif-grid'>
        <h2 className='gif-grid-title animate__animated animate__flipInX'>{category}</h2>
        <ul className='ul-gif-grid'>
            {loading ? "Loading..." : data.map( ({id,title,url}) => 
              <GifComponent 
                  key={id}
                  title={title}
                  url={url}
              />)}
        </ul>
    </div>
  )
}
