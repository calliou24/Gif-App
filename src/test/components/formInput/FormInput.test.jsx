
import jest from 'jest-mock';

import { test, expect, describe, vitest, beforeEach, isMockFunction} from 'vitest'
import { shallow } from 'enzyme' 
import { FormInput } from './../../../components/FormInput'


describe('Testing the formInput component',()=> {
    
    const  sendValue = jest.fn();
    
    let wrapper;//this requiared two functions
    
    beforeEach(()=> {
        vitest.clearAllMocks()
        wrapper = shallow(<FormInput add={sendValue} reset={sendValue}/>)  ;
    })

    test( 'The components should show well', ()=> {
        expect(wrapper).toMatchSnapshot()
    })
    
    
    test('Should simulate onChange event',()=> {
        
        const input = wrapper.find('input')
        
        const value = 'Simulated Value'
        
        input.simulate('change', { target: { value }}) //the events and the evaluations have to be on the correct steps

        const paragrapgh = wrapper.find('#p').text()
        //esto metodo simulate recibe como parametro lo que va a simular y como segundo parametro un objeto que va a ser el evento en si  
        
        expect(paragrapgh).toBe(value)
    })
    
    test('Should no execute the setCategories function', () => {

        wrapper.find('form').simulate('submit', {target : {firstChild : false}, preventDefault(){} } ) //le pasamos como segundo argumento el evento que queremos que se dispare 

        expect( sendValue ).toHaveBeenCalled()


    })
    test('Should called the sendValue and clear the input area',()=> {
        
        const form  = wrapper.find('form');
        const input  = wrapper.find('input');
        const value = 'xd';
        
        input.simulate('change', { target: { value }})
        form.simulate('submit', { target : { firstChild : true }, preventDefault(){} })
        
        const pText = wrapper.find('#p').text()
        console.log(pText)
        
        expect(  sendValue).toHaveBeenCalled()
        //expect( sendValue ).toHaveBeenCalledWith( expect.any(Function))
        expect( pText ).toBe('')

    })

})

