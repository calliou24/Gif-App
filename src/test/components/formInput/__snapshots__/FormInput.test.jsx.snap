// Vitest Snapshot v1

exports[`Testing the formInput component > The components should show well 1`] = `
<div
  className="form-cont"
>
  <p>
    Search
  </p>
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="..."
    />
    <p
      className="text-reflect"
      id="p"
    />
    <button
      className="btn-add-cat"
      type="submit"
    >
      Add
    </button>
    <button
      className="btn-add-cat"
      onClick={[Function]}
      type="submit"
    >
      Reset
    </button>
  </form>
</div>
`;
